<a href="https://colab.research.google.com/github/mehrotrasan16/Keras-Deep-Learning/blob/master/01_keras_mnist_baseline.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Things-to-try-out:" data-toc-modified-id="Things-to-try-out:-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Things to try out:</a></span></li><li><span><a href="#Model-#1-:-Number-of-Neurons" data-toc-modified-id="Model-#1-:-Number-of-Neurons-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Model #1 : Number of Neurons</a></span></li><li><span><a href="#Model-#2:-Optimizer" data-toc-modified-id="Model-#2:-Optimizer-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Model #2: Optimizer</a></span></li><li><span><a href="#Model-#3:-Adding-another-layer" data-toc-modified-id="Model-#3:-Adding-another-layer-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Model #3: Adding another layer</a></span></li><li><span><a href="#Comparison-Table" data-toc-modified-id="Comparison-Table-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Comparison Table</a></span></li></ul></div>

In [6]:
# TensorFlow and tf.keras
import tensorflow as tf
import numpy as np
from tensorflow import keras

#tf version should be 2.2 or higher
tf.__version__

'2.3.0'

In [7]:
keras.__version__

'2.4.0'

In [8]:
tf.test.is_gpu_available()

Instructions for updating:
Use `tf.config.list_physical_devices('GPU')` instead.


True

In [9]:
#get data
(train_images, train_labels), (test_images, test_labels) = \
      keras.datasets.mnist.load_data()

11493376/11490434 [==============================] - 0s 0us/step


In [10]:
#scale model
train_images = train_images / 255.0
test_images = test_images / 255.0

In [ ]:
#setup model
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(10, activation = tf.nn.softmax)
])

In [ ]:
#compile model
model.compile(optimizer='sgd',
          loss='sparse_categorical_crossentropy',
          metrics=['accuracy'])

In [ ]:
#train model
model.fit(train_images, train_labels, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 3s 1ms/step - loss: 0.7851 - accuracy: 0.8116
Epoch 2/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.4578 - accuracy: 0.8802
Epoch 3/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.4040 - accuracy: 0.8910
Epoch 4/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.3771 - accuracy: 0.8970
Epoch 5/5
1875/1875 [==============================] - 2s 1ms/step - loss: 0.3602 - accuracy: 0.9006


In [ ]:
#evaluate
test_loss, test_acc = model.evaluate(test_images,  test_labels)
print('\nTest accuracy:', test_acc)

313/313 [==============================] - 0s 967us/step - loss: 0.3359 - accuracy: 0.9079

Test accuracy: 0.9078999757766724


In [ ]:
# Predict on the first 5 test images.
predictions = model.predict(test_images[:5])

In [ ]:
#Raw predictions
predictions.shape,predictions


((5, 10),
 array([[2.21342721e-04, 1.09743360e-06, 1.71652689e-04, 2.24272953e-03,
         4.42058081e-05, 1.10166693e-04, 2.70977330e-06, 9.93551016e-01,
         2.19001129e-04, 3.43613280e-03],
        [1.58161353e-02, 3.36178142e-04, 8.30395758e-01, 3.23872976e-02,
         1.15933904e-06, 2.55346242e-02, 8.19568485e-02, 4.08923142e-07,
         1.35638425e-02, 7.79689981e-06],
        [2.75393977e-04, 9.52982366e-01, 1.50614483e-02, 6.84671942e-03,
         8.62881890e-04, 2.08819425e-03, 4.37100045e-03, 5.17000630e-03,
         1.08089820e-02, 1.53307989e-03],
        [9.96784687e-01, 1.88348590e-08, 2.67934112e-04, 6.91862369e-05,
         7.07319032e-07, 1.30594219e-03, 1.04554673e-03, 2.86947645e-04,
         1.58192124e-04, 8.07901597e-05],
        [2.44874647e-03, 4.88645674e-05, 1.10137910e-02, 9.68728971e-04,
         8.74821544e-01, 1.58484804e-03, 7.43442960e-03, 1.95053741e-02,
         1.61125790e-02, 6.60610721e-02]], dtype=float32))

In [ ]:
# Print our model's predictions
print(np.argmax(predictions, axis=1))

[7 2 1 0 4]


In [ ]:
# Check our predictions against the ground truths
print(test_labels[:5]) # [7, 2, 1, 0, 4]

[7 2 1 0 4]


# Fully Connected Layers(DNNs)
## Things to try out:
1. Number of Epochs
2. Optimizers
3. Number of Layers
4. Number of Neurons in each layer
5. Minibatch optimization

## Model #1 : Number of Neurons

**Experiment:** Increased number of epoch to 15.

**Result:** Accuracy increased from 90.78 to 91.72

In [ ]:
#setup model
model1 = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(10, activation = tf.nn.softmax)
])

In [ ]:
#compile model
model1.compile(optimizer='sgd',
          loss='sparse_categorical_crossentropy',
          metrics=['accuracy'])

In [ ]:
#train model
history = model1.fit(train_images, train_labels, epochs=15)

Epoch 1/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.7745 - accuracy: 0.8200
Epoch 2/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.4553 - accuracy: 0.8814
Epoch 3/15
1875/1875 [==============================] - 2s 992us/step - loss: 0.4029 - accuracy: 0.8918
Epoch 4/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.3768 - accuracy: 0.8974
Epoch 5/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.3601 - accuracy: 0.9013
Epoch 6/15
1875/1875 [==============================] - 2s 991us/step - loss: 0.3483 - accuracy: 0.9037
Epoch 7/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.3393 - accuracy: 0.9057
Epoch 8/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.3322 - accuracy: 0.9074
Epoch 9/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.3263 - accuracy: 0.9090
Epoch 10/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0

In [ ]:
#dir(history)
history.history

{'loss': [0.7744966745376587,
  0.4553094506263733,
  0.402904748916626,
  0.37675154209136963,
  0.360068142414093,
  0.3482522666454315,
  0.33931803703308105,
  0.33216139674186707,
  0.3263208866119385,
  0.321297824382782,
  0.31721925735473633,
  0.31356653571128845,
  0.3103366494178772,
  0.3074551820755005,
  0.30484145879745483],
 'accuracy': [0.8199833035469055,
  0.8814333081245422,
  0.8918499946594238,
  0.8973833322525024,
  0.9013166427612305,
  0.9036833047866821,
  0.9057499766349792,
  0.9074000120162964,
  0.9089999794960022,
  0.9106000065803528,
  0.9115166664123535,
  0.9131500124931335,
  0.9140166640281677,
  0.915066659450531,
  0.915316641330719]}

In [ ]:
#evaluate
test_loss, test_acc = model1.evaluate(test_images,  test_labels)
print('\nTest accuracy:', test_acc)

313/313 [==============================] - 0s 1ms/step - loss: 0.2954 - accuracy: 0.9176

Test accuracy: 0.9175999760627747


In [ ]:
# Predict on the first 5 test images.
predictions = model.predict(test_images[:5])

In [ ]:
#Raw predictions
predictions

array([[2.4456976e-04, 1.1016244e-06, 2.6870525e-04, 2.1962498e-03,
        5.0126186e-05, 1.0000253e-04, 2.9973776e-06, 9.9369860e-01,
        1.9108737e-04, 3.2465421e-03],
       [9.1871358e-03, 2.4110419e-04, 8.8870269e-01, 1.5399649e-02,
        6.8541374e-07, 1.7682318e-02, 6.0085312e-02, 3.4919714e-07,
        8.6948713e-03, 5.7744523e-06],
       [3.1419602e-04, 9.5241570e-01, 1.5809523e-02, 7.6014413e-03,
        7.0461084e-04, 2.2334328e-03, 4.3221116e-03, 5.5818781e-03,
        9.0749450e-03, 1.9421905e-03],
       [9.9548799e-01, 3.2072524e-08, 4.9561623e-04, 8.9386056e-05,
        9.7483132e-07, 1.9237131e-03, 1.1788539e-03, 5.6850811e-04,
        1.8149463e-04, 7.3461422e-05],
       [2.7817939e-03, 6.5413093e-05, 1.4476937e-02, 7.3044328e-04,
        8.8427997e-01, 1.6711350e-03, 1.0325884e-02, 2.1086464e-02,
        1.3084155e-02, 5.1497903e-02]], dtype=float32)

In [ ]:
# Print our model's predictions
print(np.argmax(predictions, axis=1))

[7 2 1 0 4]


In [ ]:
# Check our predictions against the ground truths
print(test_labels[:5]) # [7, 2, 1, 0, 4]

[7 2 1 0 4]


## Model #2: Optimizer

**Experiment:** Changed optimizer to Adam.

**Result:** Accuracy increased from 91.72 to 92.47

In [ ]:
#setup model
model2 = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(10, activation = tf.nn.softmax)
])

In [ ]:
#compile model
model2.compile(optimizer='adam',
          loss='sparse_categorical_crossentropy',
          metrics=['accuracy'])

In [ ]:
#train model
model2.fit(train_images, train_labels, epochs=15)

Epoch 1/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.4644 - accuracy: 0.8788
Epoch 2/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.3039 - accuracy: 0.9153
Epoch 3/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2837 - accuracy: 0.9211
Epoch 4/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2732 - accuracy: 0.9235
Epoch 5/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2667 - accuracy: 0.9253
Epoch 6/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2617 - accuracy: 0.9270
Epoch 7/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2581 - accuracy: 0.9281
Epoch 8/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2554 - accuracy: 0.9290
Epoch 9/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2532 - accuracy: 0.9298
Epoch 10/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.251

In [ ]:
#dir(history)
history.history

{'loss': [0.7744966745376587,
  0.4553094506263733,
  0.402904748916626,
  0.37675154209136963,
  0.360068142414093,
  0.3482522666454315,
  0.33931803703308105,
  0.33216139674186707,
  0.3263208866119385,
  0.321297824382782,
  0.31721925735473633,
  0.31356653571128845,
  0.3103366494178772,
  0.3074551820755005,
  0.30484145879745483],
 'accuracy': [0.8199833035469055,
  0.8814333081245422,
  0.8918499946594238,
  0.8973833322525024,
  0.9013166427612305,
  0.9036833047866821,
  0.9057499766349792,
  0.9074000120162964,
  0.9089999794960022,
  0.9106000065803528,
  0.9115166664123535,
  0.9131500124931335,
  0.9140166640281677,
  0.915066659450531,
  0.915316641330719]}

In [ ]:
#evaluate
test_loss, test_acc = model2.evaluate(test_images,  test_labels)
print('\nTest accuracy:', test_acc)

313/313 [==============================] - 0s 1ms/step - loss: 0.2709 - accuracy: 0.9248

Test accuracy: 0.9247999787330627


In [ ]:
# Predict on the first 5 test images.
predictions = model2.predict(test_images[:5])

In [ ]:
#Raw predictions
predictions

array([[1.41043273e-07, 3.05171307e-13, 4.72896232e-07, 6.07971940e-03,
        1.65374558e-07, 2.50882822e-05, 4.71964915e-13, 9.93602693e-01,
        1.14796703e-05, 2.80196109e-04],
       [3.72067552e-05, 1.33618153e-06, 9.93911505e-01, 1.24302378e-05,
        4.96227105e-16, 2.16920394e-03, 3.86007130e-03, 7.30927728e-21,
        8.25761981e-06, 6.21472830e-18],
       [3.56841241e-07, 9.87218380e-01, 7.58117484e-03, 1.79709191e-03,
        9.68049862e-05, 6.17539801e-04, 8.09406512e-04, 7.94487714e-05,
        1.69345306e-03, 1.06290485e-04],
       [9.99920368e-01, 2.23443312e-12, 1.22740066e-05, 6.30660452e-07,
        8.79583073e-09, 9.43566465e-06, 5.27888697e-05, 7.38457345e-07,
        1.83275267e-06, 1.95039502e-06],
       [5.31141937e-04, 4.91894383e-08, 1.15252763e-03, 1.45653466e-05,
        9.68250513e-01, 7.51500411e-05, 1.53896061e-03, 1.53930986e-03,
        1.59897143e-03, 2.52988841e-02]], dtype=float32)

In [ ]:
# Print our model's predictions
print(np.argmax(predictions, axis=1))

[7 2 1 0 4]


In [ ]:
# Check our predictions against the ground truths
print(test_labels[:5]) # [7, 2, 1, 0, 4]

[7 2 1 0 4]


## Model #3: Adding another layer

**Experiment:** Adding another Dense Layer

**Result:** Accuracy increased from 92.47 to 97.79

In [ ]:
#setup model
model3 = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation = 'relu'),
    keras.layers.Dense(10, activation = tf.nn.softmax)
])

In [ ]:
#compile model
model3.compile(optimizer='adam',
          loss='sparse_categorical_crossentropy',
          metrics=['accuracy'])

In [ ]:
#train model
model3.fit(train_images, train_labels, epochs=15)

Epoch 1/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2567 - accuracy: 0.9264
Epoch 2/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.1118 - accuracy: 0.9660
Epoch 3/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0775 - accuracy: 0.9766
Epoch 4/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0579 - accuracy: 0.9825
Epoch 5/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0447 - accuracy: 0.9863
Epoch 6/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0335 - accuracy: 0.9897
Epoch 7/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0274 - accuracy: 0.9916
Epoch 8/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0236 - accuracy: 0.9926
Epoch 9/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0172 - accuracy: 0.9949
Epoch 10/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.015

In [ ]:
#dir(history)
history.history

{'loss': [0.7744966745376587,
  0.4553094506263733,
  0.402904748916626,
  0.37675154209136963,
  0.360068142414093,
  0.3482522666454315,
  0.33931803703308105,
  0.33216139674186707,
  0.3263208866119385,
  0.321297824382782,
  0.31721925735473633,
  0.31356653571128845,
  0.3103366494178772,
  0.3074551820755005,
  0.30484145879745483],
 'accuracy': [0.8199833035469055,
  0.8814333081245422,
  0.8918499946594238,
  0.8973833322525024,
  0.9013166427612305,
  0.9036833047866821,
  0.9057499766349792,
  0.9074000120162964,
  0.9089999794960022,
  0.9106000065803528,
  0.9115166664123535,
  0.9131500124931335,
  0.9140166640281677,
  0.915066659450531,
  0.915316641330719]}

In [ ]:
#evaluate
test_loss, test_acc = model3.evaluate(test_images,  test_labels)
print('\nTest accuracy:', test_acc)

313/313 [==============================] - 0s 1ms/step - loss: 0.0970 - accuracy: 0.9779

Test accuracy: 0.9779000282287598


In [ ]:
# Predict on the first 5 test images.
predictions = model3.predict(test_images[:5])

In [ ]:
#Raw predictions
predictions

array([[2.86038832e-10, 2.45321177e-11, 3.32819272e-10, 1.37279716e-08,
        5.23444276e-18, 4.74484341e-11, 2.70333181e-13, 1.00000000e+00,
        5.40056222e-10, 2.55283386e-08],
       [1.69488854e-10, 1.19791238e-07, 9.99999881e-01, 1.85118265e-09,
        6.10233118e-28, 1.44519352e-09, 6.19506356e-12, 2.20068869e-19,
        1.14869625e-08, 3.69126723e-21],
       [2.09324669e-09, 9.99930382e-01, 2.53799726e-06, 4.45203208e-09,
        7.40185513e-08, 3.69976831e-08, 1.12523480e-07, 5.62761779e-05,
        1.06580310e-05, 1.08617748e-09],
       [1.00000000e+00, 2.67585194e-12, 2.26480862e-10, 2.28317863e-16,
        3.89202322e-14, 3.58872437e-13, 2.81044144e-09, 9.22348203e-14,
        3.63374636e-15, 2.18902424e-13],
       [2.01624099e-11, 5.06406165e-16, 7.85453091e-11, 1.13205900e-16,
        9.99999881e-01, 1.18238861e-13, 2.57706079e-11, 2.26978037e-11,
        5.65030444e-13, 9.28472161e-08]], dtype=float32)

In [ ]:
# Print our model's predictions
print(np.argmax(predictions, axis=1))

[7 2 1 0 4]


In [ ]:
# Check our predictions against the ground truths
print(test_labels[:5]) # [7, 2, 1, 0, 4]

[7 2 1 0 4]


## Model #3.1 : More Layers

**Experiment:** Adding 2 Dense Layers

**Result:** Accuracy barely increased from 97.79 to 97.86, which is possible to achieve in model #3 given a good enough start state 

In [ ]:
#setup model
model31 = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation = 'relu'),
    keras.layers.Dense(128, activation = 'relu'),
    keras.layers.Dense(10, activation = tf.nn.softmax)
])

In [ ]:
#compile model
model31.compile(optimizer='adam',
          loss='sparse_categorical_crossentropy',
          metrics=['accuracy'])

In [ ]:
#train model
model31.fit(train_images, train_labels, epochs=15)

Epoch 1/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2255 - accuracy: 0.9329
Epoch 2/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0949 - accuracy: 0.9709
Epoch 3/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0689 - accuracy: 0.9789
Epoch 4/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0520 - accuracy: 0.9833
Epoch 5/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0397 - accuracy: 0.9870
Epoch 6/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0325 - accuracy: 0.9896
Epoch 7/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0296 - accuracy: 0.9901
Epoch 8/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0222 - accuracy: 0.9926
Epoch 9/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0222 - accuracy: 0.9924
Epoch 10/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.017

In [ ]:
#dir(history)
history.history

{'loss': [0.7744966745376587,
  0.4553094506263733,
  0.402904748916626,
  0.37675154209136963,
  0.360068142414093,
  0.3482522666454315,
  0.33931803703308105,
  0.33216139674186707,
  0.3263208866119385,
  0.321297824382782,
  0.31721925735473633,
  0.31356653571128845,
  0.3103366494178772,
  0.3074551820755005,
  0.30484145879745483],
 'accuracy': [0.8199833035469055,
  0.8814333081245422,
  0.8918499946594238,
  0.8973833322525024,
  0.9013166427612305,
  0.9036833047866821,
  0.9057499766349792,
  0.9074000120162964,
  0.9089999794960022,
  0.9106000065803528,
  0.9115166664123535,
  0.9131500124931335,
  0.9140166640281677,
  0.915066659450531,
  0.915316641330719]}

In [ ]:
#evaluate
test_loss, test_acc = model31.evaluate(test_images,  test_labels)
print('\nTest accuracy:', test_acc)

313/313 [==============================] - 0s 1ms/step - loss: 0.1078 - accuracy: 0.9787

Test accuracy: 0.9786999821662903


In [ ]:
# Predict on the first 5 test images.
predictions = model31.predict(test_images[:5])

In [ ]:
#Raw predictions
predictions

array([[2.2928355e-13, 1.0564217e-08, 1.0767853e-08, 4.8878035e-10,
        6.1932474e-15, 9.8376118e-15, 3.3744430e-19, 1.0000000e+00,
        1.9511835e-11, 4.1509750e-08],
       [1.1215516e-20, 1.2560490e-13, 1.0000000e+00, 5.7819266e-13,
        9.6534242e-27, 1.2331421e-20, 6.5731358e-24, 8.4244209e-21,
        9.7911168e-14, 1.5850135e-25],
       [5.4185744e-12, 9.9998903e-01, 1.8950320e-06, 1.2045744e-09,
        1.6154329e-08, 3.7335180e-11, 5.7483369e-09, 8.7459839e-06,
        1.8242002e-07, 1.0011896e-11],
       [1.0000000e+00, 6.4671098e-12, 9.9057396e-10, 7.1977583e-16,
        2.8279757e-17, 1.2494674e-11, 7.9292251e-10, 5.7448457e-10,
        5.2910813e-16, 5.0651829e-16],
       [8.1709173e-10, 1.2238697e-14, 3.0198795e-12, 2.9030829e-19,
        9.9999845e-01, 2.9877213e-16, 1.7107191e-14, 5.0540656e-09,
        2.8402995e-11, 1.5808507e-06]], dtype=float32)

In [ ]:
# Print our model's predictions
print(np.argmax(predictions, axis=1))

[7 2 1 0 4]


In [ ]:
# Check our predictions against the ground truths
print(test_labels[:5]) # [7, 2, 1, 0, 4]

[7 2 1 0 4]


## Model #4 : Changing Learning Rate

**Experiment**: write a for loop to experiment with various learning rates

In [ ]:
#setup model
model4 = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation = 'relu'),
    keras.layers.Dense(128, activation = 'relu'),
    keras.layers.Dense(10, activation = tf.nn.softmax)
])

In [ ]:
for lr in [0.01,0.003,0.5,0.03,0.9,0.099]:
    #compile model
    model4.compile(optimizer=keras.optimizers.Adam(learning_rate=lr),
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

    #train model
    model4.fit(train_images, train_labels, epochs=15)

    #dir(history)
    history.history

    #evaluate
    test_loss, test_acc = model4.evaluate(test_images,  test_labels)
    print(f'for learning rate: {lr} : Test accuracy: {test_acc}')

Epoch 1/15
1875/1875 [==============================] - 4s 2ms/step - loss: 0.2717 - accuracy: 0.9226
Epoch 2/15
1875/1875 [==============================] - 3s 2ms/step - loss: 0.1888 - accuracy: 0.9503
Epoch 3/15
1875/1875 [==============================] - 4s 2ms/step - loss: 0.1627 - accuracy: 0.9569
Epoch 4/15
1875/1875 [==============================] - 4s 2ms/step - loss: 0.1466 - accuracy: 0.9625
Epoch 5/15
1875/1875 [==============================] - 4s 2ms/step - loss: 0.1322 - accuracy: 0.9674
Epoch 6/15
1875/1875 [==============================] - 3s 2ms/step - loss: 0.1294 - accuracy: 0.9673
Epoch 7/15
1875/1875 [==============================] - 3s 2ms/step - loss: 0.1129 - accuracy: 0.9721
Epoch 8/15
1875/1875 [==============================] - 4s 2ms/step - loss: 0.1137 - accuracy: 0.9721
Epoch 9/15
1875/1875 [==============================] - 3s 2ms/step - loss: 0.1105 - accuracy: 0.9743
Epoch 10/15
1875/1875 [==============================] - 3s 2ms/step - loss: 0.107

In [ ]:
# Predict on the first 5 test images.
predictions = model4.predict(test_images[:5])

#Raw predictions
predictions

# Print our model's predictions
print(np.argmax(predictions, axis=1))

# Check our predictions against the ground truths
print(test_labels[:5]) # [7, 2, 1, 0, 4]

[1 1 1 1 1]
[7 2 1 0 4]


Since we see that the accurace aroun 0.003 is the best, we can try exploring values in that range.

In [ ]:
#setup model
model41 = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation = 'relu'),
    keras.layers.Dense(128, activation = 'relu'),
    keras.layers.Dense(10, activation = tf.nn.softmax)
])

In [ ]:
for lr in [0.003,0.005,0.007,0.009]:
    #compile model
    model41.compile(optimizer=keras.optimizers.Adam(learning_rate=lr),
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

    #train model
    model41.fit(train_images, train_labels, epochs=15)

    #dir(history)
    history.history

    #evaluate
    test_loss, test_acc = model41.evaluate(test_images,  test_labels)
    print(f'for learning rate: {lr} : Test accuracy: {test_acc}')

Epoch 1/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2098 - accuracy: 0.9356
Epoch 2/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.1066 - accuracy: 0.9675
Epoch 3/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0822 - accuracy: 0.9749
Epoch 4/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0697 - accuracy: 0.9790
Epoch 5/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0625 - accuracy: 0.9813
Epoch 6/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0523 - accuracy: 0.9837
Epoch 7/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0471 - accuracy: 0.9858
Epoch 8/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0428 - accuracy: 0.9870
Epoch 9/15
1875/1875 [==============================] - 3s 2ms/step - loss: 0.0414 - accuracy: 0.9877
Epoch 10/15
1875/1875 [==============================] - 3s 1ms/step - loss: 0.041

Still weird, 0.003 is the best value with accuracy = 97.46

In [ ]:
# Predict on the first 5 test images.
predictions = model41.predict(test_images[:5])

#Raw predictions
predictions

# Print our model's predictions
print(np.argmax(predictions, axis=1))

# Check our predictions against the ground truths
print(test_labels[:5]) # [7, 2, 1, 0, 4]

[7 2 1 0 4]
[7 2 1 0 4]


## Model #5: Learning Rate Decay Call Back Function

In [ ]:
import math
def lr_decay(epoch):
    return 0.01 * math.pow(0.6,epoch)

lr_decay_callback = keras.callbacks.LearningRateScheduler(lr_decay,verbose=True)

In [ ]:
#setup model
model5 = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation = 'relu'),
    keras.layers.Dense(128, activation = 'relu'),
    keras.layers.Dense(10, activation = tf.nn.softmax)
])

In [ ]:
#compile model
model5.compile(optimizer='adam',
          loss='sparse_categorical_crossentropy',
          metrics=['accuracy'])

In [ ]:
#train model
model5.fit(train_images, train_labels, epochs=15,callbacks=[lr_decay_callback])


Epoch 00001: LearningRateScheduler reducing learning rate to 0.01.
Epoch 1/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2769 - accuracy: 0.9197

Epoch 00002: LearningRateScheduler reducing learning rate to 0.006.
Epoch 2/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.1369 - accuracy: 0.9610

Epoch 00003: LearningRateScheduler reducing learning rate to 0.0036.
Epoch 3/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0875 - accuracy: 0.9746

Epoch 00004: LearningRateScheduler reducing learning rate to 0.0021599999999999996.
Epoch 4/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.0559 - accuracy: 0.9827

Epoch 00005: LearningRateScheduler reducing learning rate to 0.001296.
Epoch 5/15
1875/1875 [==============================] - 3s 1ms/step - loss: 0.0342 - accuracy: 0.9893

Epoch 00006: LearningRateScheduler reducing learning rate to 0.0007775999999999998.
Epoch 6/15
1875/1875 [===================

In [ ]:
#dir(history)
history.history

{'loss': [0.7744966745376587,
  0.4553094506263733,
  0.402904748916626,
  0.37675154209136963,
  0.360068142414093,
  0.3482522666454315,
  0.33931803703308105,
  0.33216139674186707,
  0.3263208866119385,
  0.321297824382782,
  0.31721925735473633,
  0.31356653571128845,
  0.3103366494178772,
  0.3074551820755005,
  0.30484145879745483],
 'accuracy': [0.8199833035469055,
  0.8814333081245422,
  0.8918499946594238,
  0.8973833322525024,
  0.9013166427612305,
  0.9036833047866821,
  0.9057499766349792,
  0.9074000120162964,
  0.9089999794960022,
  0.9106000065803528,
  0.9115166664123535,
  0.9131500124931335,
  0.9140166640281677,
  0.915066659450531,
  0.915316641330719]}

In [ ]:
#evaluate
test_loss, test_acc = model5.evaluate(test_images,  test_labels)
print('\nTest accuracy:', test_acc)

313/313 [==============================] - 1s 2ms/step - loss: 0.1070 - accuracy: 0.9786

Test accuracy: 0.978600025177002


In [ ]:
# Predict on the first 5 test images.
predictions = model5.predict(test_images[:5])

In [ ]:
#Raw predictions
predictions

array([[2.2928355e-13, 1.0564217e-08, 1.0767853e-08, 4.8878035e-10,
        6.1932474e-15, 9.8376118e-15, 3.3744430e-19, 1.0000000e+00,
        1.9511835e-11, 4.1509750e-08],
       [1.1215516e-20, 1.2560490e-13, 1.0000000e+00, 5.7819266e-13,
        9.6534242e-27, 1.2331421e-20, 6.5731358e-24, 8.4244209e-21,
        9.7911168e-14, 1.5850135e-25],
       [5.4185744e-12, 9.9998903e-01, 1.8950320e-06, 1.2045744e-09,
        1.6154329e-08, 3.7335180e-11, 5.7483369e-09, 8.7459839e-06,
        1.8242002e-07, 1.0011896e-11],
       [1.0000000e+00, 6.4671098e-12, 9.9057396e-10, 7.1977583e-16,
        2.8279757e-17, 1.2494674e-11, 7.9292251e-10, 5.7448457e-10,
        5.2910813e-16, 5.0651829e-16],
       [8.1709173e-10, 1.2238697e-14, 3.0198795e-12, 2.9030829e-19,
        9.9999845e-01, 2.9877213e-16, 1.7107191e-14, 5.0540656e-09,
        2.8402995e-11, 1.5808507e-06]], dtype=float32)

In [ ]:
# Print our model's predictions
print(np.argmax(predictions, axis=1))

[7 2 1 0 4]


In [ ]:
# Check our predictions against the ground truths
print(test_labels[:5]) # [7, 2, 1, 0, 4]

[7 2 1 0 4]


## Model #6 : Adding dropout layers

This didn't increase the performance much higher than our 2 layer network, now let us try adding drop out layers

In [ ]:
import math
def lr_decay(epoch):
    return 0.01 * math.pow(0.6,epoch)

lr_decay_callback = keras.callbacks.LearningRateScheduler(lr_decay,verbose=True)

In [ ]:
#setup model
model6 = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation = 'relu'),
    keras.layers.Dropout(0.25),
    keras.layers.Dense(128, activation = 'relu'),
    keras.layers.Dropout(0.25),
    keras.layers.Dense(10, activation = tf.nn.softmax)
])

In [ ]:
#compile model
model6.compile(optimizer='adam',
          loss='sparse_categorical_crossentropy',
          metrics=['accuracy'])

In [ ]:
#train model
model6.fit(train_images, train_labels, epochs=15,callbacks=[lr_decay_callback])


Epoch 00001: LearningRateScheduler reducing learning rate to 0.01.
Epoch 1/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.4744 - accuracy: 0.8667

Epoch 00002: LearningRateScheduler reducing learning rate to 0.006.
Epoch 2/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2991 - accuracy: 0.9189

Epoch 00003: LearningRateScheduler reducing learning rate to 0.0036.
Epoch 3/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.2303 - accuracy: 0.9354

Epoch 00004: LearningRateScheduler reducing learning rate to 0.0021599999999999996.
Epoch 4/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.1881 - accuracy: 0.9472

Epoch 00005: LearningRateScheduler reducing learning rate to 0.001296.
Epoch 5/15
1875/1875 [==============================] - 2s 1ms/step - loss: 0.1614 - accuracy: 0.9535

Epoch 00006: LearningRateScheduler reducing learning rate to 0.0007775999999999998.
Epoch 6/15
1875/1875 [===================

In [ ]:
#dir(history)
history.history

{'loss': [0.7744966745376587,
  0.4553094506263733,
  0.402904748916626,
  0.37675154209136963,
  0.360068142414093,
  0.3482522666454315,
  0.33931803703308105,
  0.33216139674186707,
  0.3263208866119385,
  0.321297824382782,
  0.31721925735473633,
  0.31356653571128845,
  0.3103366494178772,
  0.3074551820755005,
  0.30484145879745483],
 'accuracy': [0.8199833035469055,
  0.8814333081245422,
  0.8918499946594238,
  0.8973833322525024,
  0.9013166427612305,
  0.9036833047866821,
  0.9057499766349792,
  0.9074000120162964,
  0.9089999794960022,
  0.9106000065803528,
  0.9115166664123535,
  0.9131500124931335,
  0.9140166640281677,
  0.915066659450531,
  0.915316641330719]}

In [ ]:
#evaluate
test_loss, test_acc = model5.evaluate(test_images,  test_labels)
print('\nTest accuracy:', test_acc)

313/313 [==============================] - 1s 2ms/step - loss: 0.1070 - accuracy: 0.9786

Test accuracy: 0.978600025177002


In [ ]:
# Predict on the first 5 test images.
predictions = model6.predict(test_images[:5])

In [ ]:
#Raw predictions
predictions

array([[6.3150554e-13, 2.4647166e-11, 1.0940240e-08, 9.6726609e-09,
        1.4008019e-09, 4.9387622e-11, 5.3494877e-21, 9.9999809e-01,
        7.2885127e-15, 1.8830224e-06],
       [1.2051256e-09, 7.1458978e-04, 9.9917644e-01, 8.9442328e-05,
        2.5940569e-10, 1.9257165e-08, 1.2127187e-09, 1.8187713e-05,
        1.3230484e-06, 3.8495303e-13],
       [7.9323017e-19, 9.9999952e-01, 2.1289546e-13, 1.3772233e-12,
        1.9057420e-09, 2.0326343e-12, 3.6850570e-12, 4.2401754e-07,
        3.4652398e-08, 4.5222489e-12],
       [9.9998438e-01, 2.6231689e-20, 2.1760754e-07, 9.1002885e-09,
        8.0106606e-09, 3.3857009e-06, 5.0838380e-06, 4.4168247e-09,
        1.9793354e-09, 6.8763211e-06],
       [5.7425915e-08, 3.1153922e-07, 2.4018238e-06, 3.4222283e-09,
        9.9978966e-01, 1.3571064e-07, 3.0250439e-07, 1.9755407e-05,
        9.0138137e-08, 1.8736135e-04]], dtype=float32)

In [ ]:
# Print our model's predictions
print(np.argmax(predictions, axis=1))

[7 2 1 0 4]


In [ ]:
# Check our predictions against the ground truths
print(test_labels[:5]) # [7, 2, 1, 0, 4]

[7 2 1 0 4]


## Comparison Table

| Approach | Accuracy(%) |
|---|---|
| Base Model | 90.78 |
| Epoch inc to 15| 91.72 |
| ADAM Optimizer | 92.65 |
| 128 - Dense layer | 97.79 |
| 2 128 - Dense layers | 97.87   |
| Best of learning rate loop: 0.003 | 97.44  |
| with Learning Rate Decay | 97.86  |
| with Dropout layers | 97.86  |

**Conclusion**:
With Fully Connected DNNs, we seem to hit a cap of 97.86 no matter what we do. To get this to 99.3% we must incolve Conv2D , CNNs with Dropout layers. 

# Convolutional Neural Networks 

In [11]:
train_images = train_images.reshape(train_images.shape[0],28,28,1)
test_images = test_images.reshape(test_images.shape[0],28,28,1)

In [ ]:
import math
def lr_decay(epoch):
    return 0.01 * math.pow(0.6,epoch)

lr_decay_callback = keras.callbacks.LearningRateScheduler(lr_decay,verbose=True)

# 98.48

In [ ]:
#setup model
cnnmodel1 = keras.Sequential([
    keras.layers.Conv2D(kernel_size=3,filters=12,activation='relu',padding='same',input_shape=(28,28,1)),
    keras.layers.MaxPooling2D(pool_size=(2,2)),
    keras.layers.Conv2D(kernel_size=6,filters=24,activation='relu',padding='same',strides=3),
    keras.layers.MaxPooling2D(pool_size=(2,2)),
    keras.layers.Conv2D(kernel_size=6,filters=32,activation='relu',padding='same',strides=3),
    keras.layers.Flatten(),
    keras.layers.Dense(128, activation='relu'),
    keras.layers.Dense(10, activation = tf.nn.softmax)    
])

In [ ]:
#compile model
cnnmodel1.compile(optimizer='adam',
          loss='sparse_categorical_crossentropy',
          metrics=['accuracy'])

In [ ]:
#train model
cnnmodel1.fit(train_images, train_labels, epochs=15,callbacks=[lr_decay_callback])


Epoch 00001: LearningRateScheduler reducing learning rate to 0.01.
Epoch 1/15
1875/1875 [==============================] - 32s 17ms/step - loss: 0.2312 - accuracy: 0.9286

Epoch 00002: LearningRateScheduler reducing learning rate to 0.006.
Epoch 2/15
1875/1875 [==============================] - 33s 18ms/step - loss: 0.0968 - accuracy: 0.9715

Epoch 00003: LearningRateScheduler reducing learning rate to 0.0036.
Epoch 3/15
1875/1875 [==============================] - 33s 18ms/step - loss: 0.0633 - accuracy: 0.9817

Epoch 00004: LearningRateScheduler reducing learning rate to 0.0021599999999999996.
Epoch 4/15
1875/1875 [==============================] - 34s 18ms/step - loss: 0.0426 - accuracy: 0.9871

Epoch 00005: LearningRateScheduler reducing learning rate to 0.001296.
Epoch 5/15
1875/1875 [==============================] - 33s 18ms/step - loss: 0.0321 - accuracy: 0.9898

Epoch 00006: LearningRateScheduler reducing learning rate to 0.0007775999999999998.
Epoch 6/15
1875/1875 [=========

In [ ]:
#evaluate
test_loss, test_acc = cnnmodel1.evaluate(test_images,  test_labels)
print('\nTest accuracy:', test_acc)

313/313 [==============================] - 3s 8ms/step - loss: 0.0674 - accuracy: 0.9849

Test accuracy: 0.9848999977111816


In [ ]:
# Predict on the first 5 test images.
predictions = cnnmodel1.predict(test_images[:5])

In [ ]:
#Raw predictions
predictions

array([[2.58821339e-13, 4.42130360e-10, 2.81399093e-06, 8.22663992e-10,
        4.22725073e-08, 1.46156628e-10, 3.79519306e-19, 9.99997020e-01,
        1.57904023e-10, 1.16600425e-07],
       [1.49077127e-12, 1.85843983e-12, 1.00000000e+00, 4.35442698e-12,
        2.57657827e-12, 6.03696712e-14, 1.64096066e-14, 5.49620523e-11,
        2.45187880e-11, 8.77612372e-16],
       [1.45968844e-16, 1.00000000e+00, 4.10229779e-14, 9.53375804e-19,
        3.30425687e-11, 5.42400597e-14, 4.69045679e-12, 2.23021471e-14,
        2.85816596e-12, 3.74989861e-13],
       [9.99956965e-01, 1.10112047e-14, 2.14770146e-10, 2.89034602e-10,
        1.29831657e-09, 1.60293254e-08, 3.27569251e-05, 1.62169421e-12,
        5.78023451e-09, 1.02208933e-05],
       [3.10790182e-18, 1.15170070e-14, 1.86831209e-13, 1.04194933e-20,
        1.00000000e+00, 8.31934267e-16, 6.68590888e-14, 4.18914306e-13,
        2.64455992e-13, 8.70124042e-11]], dtype=float32)

In [ ]:
# Print our model's predictions
print(np.argmax(predictions, axis=1))

[7 2 1 0 4]


In [ ]:
# Check our predictions against the ground truths
print(test_labels[:5]) # [7, 2, 1, 0, 4]

[7 2 1 0 4]


# 98.79

In [ ]:
#setup model
cnnmodel2 = keras.Sequential([
    keras.layers.Conv2D(kernel_size=3,filters=12,activation='relu',padding='same',input_shape=(28,28,1)),
    keras.layers.Conv2D(kernel_size=6,filters=24,activation='relu',padding='same',strides=3),
    keras.layers.Conv2D(kernel_size=6,filters=32,activation='relu',padding='same',strides=3),
    keras.layers.Flatten(),
    keras.layers.Dense(128, activation='relu'),
    #keras.layers.Dropout(0.25),
    keras.layers.Dense(10, activation = tf.nn.softmax)    
])

In [ ]:
#compile model
cnnmodel2.compile(optimizer='adam',
          loss='sparse_categorical_crossentropy',
          metrics=['accuracy'])

In [ ]:
#train model
cnnmodel2.fit(train_images, train_labels, epochs=10)

Epoch 1/10
1875/1875 [==============================] - 49s 26ms/step - loss: 0.1642 - accuracy: 0.9486
Epoch 2/10
1875/1875 [==============================] - 49s 26ms/step - loss: 0.0553 - accuracy: 0.9828
Epoch 3/10
1875/1875 [==============================] - 52s 28ms/step - loss: 0.0374 - accuracy: 0.9885
Epoch 4/10
1875/1875 [==============================] - 50s 27ms/step - loss: 0.0278 - accuracy: 0.9911
Epoch 5/10
1875/1875 [==============================] - 50s 27ms/step - loss: 0.0212 - accuracy: 0.9929
Epoch 6/10
1875/1875 [==============================] - 51s 27ms/step - loss: 0.0169 - accuracy: 0.9948
Epoch 7/10
1875/1875 [==============================] - 50s 27ms/step - loss: 0.0137 - accuracy: 0.9955
Epoch 8/10
1875/1875 [==============================] - 50s 27ms/step - loss: 0.0129 - accuracy: 0.9957
Epoch 9/10
1875/1875 [==============================] - 51s 27ms/step - loss: 0.0106 - accuracy: 0.9966
Epoch 10/10
1875/1875 [==============================] - 49s 26m

In [ ]:
#evaluate
test_loss, test_acc = cnnmodel2.evaluate(test_images,  test_labels)
print('\nTest accuracy:', test_acc)

313/313 [==============================] - 3s 10ms/step - loss: 0.0452 - accuracy: 0.9879

Test accuracy: 0.9879000186920166


In [ ]:
# Predict on the first 5 test images.
predictions = cnnmodel2.predict(test_images[:5])

In [ ]:
#Raw predictions
predictions

array([[9.1748900e-12, 9.1796556e-11, 2.7299302e-08, 3.6320422e-11,
        9.8894764e-12, 4.1397853e-13, 2.6194843e-17, 1.0000000e+00,
        6.2747745e-13, 2.2187985e-09],
       [7.1729658e-20, 1.1877107e-15, 1.0000000e+00, 1.1910056e-21,
        4.0960795e-23, 6.2814627e-26, 1.1696030e-18, 4.7109088e-23,
        8.9971780e-15, 2.7189868e-23],
       [4.1285803e-06, 9.9987817e-01, 3.0863303e-06, 2.2479698e-09,
        1.4219809e-06, 2.9214144e-07, 1.5567189e-06, 1.2510365e-06,
        1.1002855e-04, 1.6997017e-08],
       [9.9999952e-01, 1.8046469e-17, 2.3325680e-10, 7.8363465e-15,
        1.0223754e-11, 4.1493212e-11, 2.2690299e-08, 4.9674110e-12,
        2.4817688e-11, 4.6448841e-07],
       [2.7687352e-11, 4.4522434e-11, 2.8588606e-10, 8.1518204e-12,
        9.9994826e-01, 3.4906242e-10, 8.7942237e-11, 1.3688676e-11,
        1.2679978e-09, 5.1759605e-05]], dtype=float32)

In [ ]:
# Print our model's predictions
print(np.argmax(predictions, axis=1))

[7 2 1 0 4]


In [ ]:
# Check our predictions against the ground truths
print(test_labels[:5]) # [7, 2, 1, 0, 4]

[7 2 1 0 4]


# 98.94

In [39]:
#setup model
cnnmodel3 = keras.Sequential([
    keras.layers.Conv2D(kernel_size=3,filters=12,activation='relu',padding='same',input_shape=(28,28,1)),
    keras.layers.Conv2D(kernel_size=6,filters=24,activation='relu',padding='same',strides=3),
    keras.layers.Conv2D(kernel_size=6,filters=32,activation='relu',padding='same',strides=3),
    keras.layers.Flatten(),
    keras.layers.Dense(128, activation='relu'),
    keras.layers.Dropout(0.4),
    keras.layers.Dense(200, activation='relu'),
    keras.layers.Dropout(0.4),
    keras.layers.Dense(10, activation = tf.nn.softmax)    
])

In [40]:
#compile model
cnnmodel3.compile(optimizer='adam',
          loss='sparse_categorical_crossentropy',
          metrics=['accuracy'])

In [41]:
#train model
cnnmodel3.fit(train_images, train_labels, epochs=10)

Epoch 1/10
1875/1875 [==============================] - 53s 28ms/step - loss: 0.2596 - accuracy: 0.9220
Epoch 2/10
1875/1875 [==============================] - 54s 29ms/step - loss: 0.0893 - accuracy: 0.9750
Epoch 3/10
1875/1875 [==============================] - 51s 27ms/step - loss: 0.0631 - accuracy: 0.9826
Epoch 4/10
1875/1875 [==============================] - 50s 27ms/step - loss: 0.0502 - accuracy: 0.9859
Epoch 5/10
1875/1875 [==============================] - 52s 28ms/step - loss: 0.0428 - accuracy: 0.9883
Epoch 6/10
1875/1875 [==============================] - 51s 27ms/step - loss: 0.0353 - accuracy: 0.9899
Epoch 7/10
1875/1875 [==============================] - 51s 27ms/step - loss: 0.0299 - accuracy: 0.9913
Epoch 8/10
1875/1875 [==============================] - 52s 28ms/step - loss: 0.0280 - accuracy: 0.9918
Epoch 9/10
1875/1875 [==============================] - 52s 28ms/step - loss: 0.0260 - accuracy: 0.9924
Epoch 10/10
1875/1875 [==============================] - 51s 27m

In [42]:
#evaluate
test_loss, test_acc = cnnmodel3.evaluate(test_images,  test_labels)
print('\nTest accuracy:', test_acc)

313/313 [==============================] - 3s 10ms/step - loss: 0.0458 - accuracy: 0.9895

Test accuracy: 0.9894999861717224


In [ ]:
# Predict on the first 5 test images.
predictions = cnnmodel3.predict(test_images[:5])

In [ ]:
#Raw predictions
predictions

array([[9.1748900e-12, 9.1796556e-11, 2.7299302e-08, 3.6320422e-11,
        9.8894764e-12, 4.1397853e-13, 2.6194843e-17, 1.0000000e+00,
        6.2747745e-13, 2.2187985e-09],
       [7.1729658e-20, 1.1877107e-15, 1.0000000e+00, 1.1910056e-21,
        4.0960795e-23, 6.2814627e-26, 1.1696030e-18, 4.7109088e-23,
        8.9971780e-15, 2.7189868e-23],
       [4.1285803e-06, 9.9987817e-01, 3.0863303e-06, 2.2479698e-09,
        1.4219809e-06, 2.9214144e-07, 1.5567189e-06, 1.2510365e-06,
        1.1002855e-04, 1.6997017e-08],
       [9.9999952e-01, 1.8046469e-17, 2.3325680e-10, 7.8363465e-15,
        1.0223754e-11, 4.1493212e-11, 2.2690299e-08, 4.9674110e-12,
        2.4817688e-11, 4.6448841e-07],
       [2.7687352e-11, 4.4522434e-11, 2.8588606e-10, 8.1518204e-12,
        9.9994826e-01, 3.4906242e-10, 8.7942237e-11, 1.3688676e-11,
        1.2679978e-09, 5.1759605e-05]], dtype=float32)

In [ ]:
# Print our model's predictions
print(np.argmax(predictions, axis=1))

[7 2 1 0 4]


In [ ]:
# Check our predictions against the ground truths
print(test_labels[:5]) # [7, 2, 1, 0, 4]

[7 2 1 0 4]


# 

In [13]:
#setup model
cnnmodel4 = keras.Sequential([
    keras.layers.Conv2D(kernel_size=3,filters=12,use_bias=False,padding='same'),
    keras.layers.BatchNormalization(center=True,scale=False),
    keras.layers.Activation('relu'),

    keras.layers.Conv2D(kernel_size=6,filters=24,use_bias=False,padding='same',strides=2),
    keras.layers.BatchNormalization(center=True,scale=False),
    keras.layers.Activation('relu'),
    
    keras.layers.Conv2D(kernel_size=6,filters=32,use_bias=False,padding='same',strides=2),
    keras.layers.BatchNormalization(center=True,scale=False),
    keras.layers.Activation('relu'),
    
    keras.layers.Flatten(),
    
    keras.layers.Dense(128),
    keras.layers.BatchNormalization(center=True,scale=False),
    keras.layers.Activation('relu'),
    keras.layers.Dropout(0.4),

    keras.layers.Dense(200),
    keras.layers.BatchNormalization(center=True,scale=False),
    keras.layers.Activation('relu'),
    keras.layers.Dropout(0.4),
    
    keras.layers.Dense(10, activation = tf.nn.softmax)    
])

In [14]:
#compile model
cnnmodel4.compile(optimizer='adam',
          loss='sparse_categorical_crossentropy',
          metrics=['accuracy'])

In [15]:
#train model
cnnmodel4.fit(train_images, train_labels, epochs=10)

Epoch 1/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.2169 - accuracy: 0.9351
Epoch 2/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0809 - accuracy: 0.9759
Epoch 3/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0626 - accuracy: 0.9819
Epoch 4/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0523 - accuracy: 0.9848
Epoch 5/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0430 - accuracy: 0.9873
Epoch 6/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0387 - accuracy: 0.9880
Epoch 7/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0348 - accuracy: 0.9896
Epoch 8/10
1875/1875 [==============================] - 6s 3ms/step - loss: 0.0293 - accuracy: 0.9910
Epoch 9/10
1875/1875 [==============================] - 7s 3ms/step - loss: 0.0264 - accuracy: 0.9917
Epoch 10/10
1875/1875 [==============================] - 7s 4ms/step - loss: 0.023

In [17]:
#evaluate
test_loss, test_acc = cnnmodel4.evaluate(test_images,  test_labels)
print('\nTest accuracy:', test_acc)

313/313 [==============================] - 1s 2ms/step - loss: 0.0217 - accuracy: 0.9930

Test accuracy: 0.9929999709129333


In [18]:
# Predict on the first 5 test images.
predictions = cnnmodel4.predict(test_images[:5])

In [19]:
#Raw predictions
predictions

array([[8.15029988e-10, 6.32423962e-08, 1.49575797e-07, 4.90447150e-09,
        4.88744085e-08, 4.83876716e-10, 5.69438852e-09, 9.99999166e-01,
        8.34515027e-11, 6.13133807e-07],
       [1.54924024e-07, 3.56493821e-08, 9.99994516e-01, 1.02682804e-07,
        3.28987397e-08, 4.01041570e-12, 4.93736297e-06, 1.99132330e-07,
        1.24171269e-08, 5.16413845e-09],
       [3.17750465e-10, 1.00000000e+00, 1.05235687e-09, 1.21815127e-11,
        2.25314503e-10, 7.56320145e-11, 7.45349737e-09, 1.44219303e-08,
        1.30716185e-08, 2.39522256e-11],
       [9.99992847e-01, 5.15069551e-08, 9.99725547e-09, 2.39075315e-09,
        4.26711289e-08, 3.13056091e-07, 4.29830470e-06, 1.55122905e-07,
        1.62151912e-06, 7.66250992e-07],
       [1.31871081e-09, 2.82378380e-07, 5.03988240e-10, 1.55313193e-11,
        9.99982953e-01, 7.10562553e-10, 8.89375773e-09, 4.94240426e-09,
        1.31498590e-09, 1.68008792e-05]], dtype=float32)

In [20]:
# Print our model's predictions
print(np.argmax(predictions, axis=1))

[7 2 1 0 4]


In [21]:
# Check our predictions against the ground truths
print(test_labels[:5]) # [7, 2, 1, 0, 4]

[7 2 1 0 4]
